common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+              type: 'servicebus.windows.net:sastoken',
+              name: audience
+            };
+            amqpMessage.body = token;
+            amqpMessage.properties = {
+              to: '$cbs',
+              messageId: uuid.v4(),
+              reply_to: 'cbs'
+            };
+
+            let outstandingPutToken: PutTokenOperation = {
+              putTokenCallback: putTokenCallback,
+              expirationTime: Math.round(Date.now() / 1000) + this._putToken.numberOfSecondsToTimeout,
+              correlationId: amqpMessage.properties.messageId
+            };
+
+            this._putToken.outstandingPutTokens.push(outstandingPutToken);
+            //
+            // If this is the first put token then start trying to time it out.
+            //
+            if (this._putToken.outstandingPutTokens.length === 1) {
+              this._putToken.timeoutTimer = setTimeout(this._removeExpiredPutTokens.bind(this), this._putToken.putTokenTimeOutExaminationInterval);
+            }
+            /*Codes_SRS_NODE_AMQP_CBS_16_012: [The `putToken` method shall send this message over the `$cbs` sender link.]*/
+            this._senderLink.send(amqpMessage, (err) => {
+              if (err) {
+                // Find the operation in the outstanding array.  Remove it from the array since, well, it's not outstanding anymore.
+                // Since we may have arrived here asynchronously, we simply can't assume that it is the end of the array.  But,
+                // it's more likely near the end.
+                for (let i = this._putToken.outstandingPutTokens.length - 1; i >= 0; i--) {
+                  if (this._putToken.outstandingPutTokens[i].correlationId === amqpMessage.properties.messageId) {
+                    const outStandingPutTokenInError = this._putToken.outstandingPutTokens[i];
+                    this._putToken.outstandingPutTokens.splice(i, 1);
+                    //
+                    // This was the last outstanding put token.  No point in having a timer around trying to time nothing out.
+                    //
+                    if (this._putToken.outstandingPutTokens.length === 0) {
+                      clearTimeout(this._putToken.timeoutTimer);
+                    }
+                    /*Codes_SRS_NODE_AMQP_CBS_16_013: [The `putToken` method shall call `callback` (if supplied) if the `send` generates an error such that no response from the service will be forthcoming.]*/
+                    outStandingPutTokenInError.putTokenCallback(err);
+                    break;
+                  }
+                }
+              }
+            });
+          }
+        },
+        detaching: {
+          _onEnter: (forwardedCallback, err) => {
+            /*Codes_SRS_NODE_AMQP_CBS_16_008: [`detach` shall detach both sender and receiver links and return the state machine to the `detached` state.]*/
+            this._senderLink.detach(() => {
+              this._receiverLink.detach(() => {
```

Comment:
Should you handle null _senderLink or _receiverLink?  I don 't find any easy failures to get into this state, but this could result in an exception or broken state.